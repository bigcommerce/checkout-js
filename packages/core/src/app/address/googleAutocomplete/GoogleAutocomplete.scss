@import '../../ui/Base';

$googleLogoHeight: 18px;

.co-googleAutocomplete-footer {
    background-color: color("greys", "lightest");
    background-image: url("../../../static/img/powered_by_google_on_white.png");
    background-position-x: spacing("half");
    background-position-y: center;
    background-repeat: no-repeat;
    border-top: container("border");
    box-sizing: content-box;
    color: color("greys", "dark");
    height: $googleLogoHeight;
    padding: spacing("half");
}
