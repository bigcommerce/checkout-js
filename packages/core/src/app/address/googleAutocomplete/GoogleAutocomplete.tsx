import { noop } from 'lodash';
import React, { PureComponent, ReactNode } from 'react';

import { Autocomplete, AutocompleteItem } from '../../ui/autocomplete';

import { GoogleAutocompleteOptionTypes } from './googleAutocompleteTypes';
import './GoogleAutocomplete.scss';
import GoogleAutocompleteService from './GoogleAutocompleteService';

interface GoogleAutocompleteProps {
    initialValue?: string;
    componentRestrictions?: google.maps.places.ComponentRestrictions;
    fields?: string[];
    apiKey: string;
    nextElement?: HTMLElement;
    inputProps?: any;
    isAutocompleteEnabled?: boolean;
    types?: GoogleAutocompleteOptionTypes[];
    onSelect?(place: google.maps.places.PlaceResult, item: AutocompleteItem): void;
    onToggleOpen?(state: { inputValue: string; isOpen: boolean }): void;
    onChange?(value: string, isOpen: boolean): void;
}

interface GoogleAutocompleteState {
    items: AutocompleteItem[];
    autoComplete: string;
}

class GoogleAutocomplete extends PureComponent<GoogleAutocompleteProps, GoogleAutocompleteState> {
    googleAutocompleteService: GoogleAutocompleteService;

    constructor(props: GoogleAutocompleteProps) {
        super(props);
        this.googleAutocompleteService = new GoogleAutocompleteService(props.apiKey);
        this.state = {
            items: [],
            autoComplete: 'off',
        };
    }

    render(): ReactNode {
        const { initialValue, onToggleOpen = noop, inputProps = {} } = this.props;

        const { autoComplete, items } = this.state;

        return (
            <Autocomplete
                initialHighlightedIndex={0}
                initialValue={initialValue}
                inputProps={{
                    ...inputProps,
                    autoComplete,
                }}
                items={items}
                listTestId="address-autocomplete-suggestions"
                onChange={this.onChange}
                onSelect={this.onSelect}
                onToggleOpen={onToggleOpen}
            >
                <div className="co-googleAutocomplete-footer" />
            </Autocomplete>
        );
    }

    private onSelect: (item: AutocompleteItem) => void = (item) => {
        const { fields, onSelect = noop, nextElement } = this.props;

        this.googleAutocompleteService.getPlacesServices().then((service) => {
            service.getDetails(
                {
                    placeId: item.id,
                    fields: fields || ['address_components', 'name'],
                },
                (result) => {
                    if (nextElement) {
                        nextElement.focus();
                    }

                    onSelect(result, item);
                },
            );
        });
    };

    private onChange: (input: string) => void = (input) => {
        const { isAutocompleteEnabled, onChange = noop } = this.props;

        onChange(input, false);

        if (!isAutocompleteEnabled) {
            return this.resetAutocomplete();
        }

        this.setAutocomplete(input);
        this.setItems(input);
    };

    private setItems(input: string): void {
        if (!input) {
            this.setState({ items: [] });

            return;
        }

        const { componentRestrictions, types } = this.props;

        this.googleAutocompleteService.getAutocompleteService().then((service) => {
            service.getPlacePredictions(
                {
                    input,
                    types: types || ['geocode'],
                    componentRestrictions,
                },
                (results) => this.setState({ items: this.toAutocompleteItems(results) }),
            );
        });
    }

    private resetAutocomplete(): void {
        this.setState({
            items: [],
            autoComplete: 'off',
        });
    }

    private setAutocomplete(input: string): void {
        this.setState({
            ...this.state,
            autoComplete: input && input.length ? 'nope' : 'off',
        });
    }

    private toAutocompleteItems(
        results?: google.maps.places.AutocompletePrediction[],
    ): AutocompleteItem[] {
        return (results || []).map((result) => ({
            label: result.description,
            value: result.structured_formatting.main_text,
            highlightedSlices: result.matched_substrings,
            id: result.place_id,
        }));
    }
}

export default GoogleAutocomplete;
