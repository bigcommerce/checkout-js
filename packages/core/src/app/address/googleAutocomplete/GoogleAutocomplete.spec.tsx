import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import GoogleAutocomplete from './GoogleAutocomplete';

describe('GoogleAutocomplete Component', () => {
    // @todo: add more tests when we can successfully mock "GoogleAutocompleteService"
    // At the moment, jest.mock('...') doesn't seem to do the trick.

    it('renders input with initial value', () => {
        const tree = render(
            <GoogleAutocomplete
                apiKey="bar"
                initialValue="fo"
                inputProps={{ placeholder: 'NO PO BOX' }}
            />,
        );

        expect(toJson(tree)).toMatchSnapshot();
    });
});
