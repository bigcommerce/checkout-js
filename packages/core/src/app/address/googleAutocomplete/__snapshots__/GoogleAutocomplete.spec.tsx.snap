// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleAutocomplete Component renders input with initial value 1`] = `
<div
  aria-expanded="false"
  aria-haspopup="listbox"
  aria-labelledby="downshift-0-label"
  role="combobox"
>
  <input
    aria-autocomplete="list"
    aria-labelledby="downshift-0-label"
    autocomplete="off"
    id="downshift-0-input"
    placeholder="NO PO BOX"
    value="fo"
  />
</div>
`;
